lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -57,140 +61,161 @@ def tags(repository, params = {})
+          unauthenticated_repos | user.repositories_dataset.select(:name)
         end
       end
 
       def unauthenticated_repos
-        Repository.where(auth_required: false).order(:name).map(:name)
+        #database = initialize_db
+        database[:repositories].where(auth_required: false).order(:name).select(:name)
       end
 
       # Replaces the entire list of repositories
       def update_repository_list(repo_list)
-        RepositoryUser.dataset.delete
-        Repository.dataset.delete
-        repo_list.each do |repo|
-          Repository.find_or_create(name: repo['repository'],
-                                    auth_required: repo['auth_required'].to_s.downcase == "true")
-        end
+        # TODO: make repositories_users cascade on delete repository
+        #database = initialize_db
+        database[:repositories_users].delete
+
+        repository = database[:repositories]
+        repository.delete
+
+        repository.import(
+          [:name, :auth_required],
+          repo_list.map { |repo| [repo['repository'], repo['auth_required'].to_s.downcase == "true"] },
+        )
       end
 
       # Replaces the entire user-repo mapping for all logged-in users
       def update_user_repo_mapping(user_repo_maps)
         # Get hash map of all users and their repositories
         # Ex: {"users"=> [{"admin"=>[{"repository"=>"repo", "auth_required"=>"true"}]}]}
         # Go through list of repositories and add them to the DB
-        RepositoryUser.dataset.delete
-        user_repo_maps['users'].each do |user_repo_map|
-          user_repo_map.each do |user, repos|
-            next if repos.nil?
-
-            repos.each do |repo|
-              found_repo = Repository.find(name: repo['repository'],
-                                           auth_required: repo['auth_required'].to_s.downcase == "true")
-              if found_repo.nil?
-                logger.warn("#{repo['repository']} does not exist in this smart proxy's environments")
-              elsif found_repo.auth_required
-                found_repo.add_user(User.find(name: user))
-              end
-            end
+        #database = initialize_db
+        repositories = database[:repositories]
+
+        entries = user_repo_maps['users'].flat_map do |user_repo_map|
+          user_repo_map.filter_map do |username, repos|
+            user_repo_names = repos.filter { |repo| repo['auth_required'].to_s.downcase == "true" }.map { |repo| repo['repository'] }
+            user = database[:users][{name: username}]
+            repositories.where(name: user_repo_names, auth_required: true).select(:id).map { |repo| [repo[:id], user[:id]] }
           end
         end
+
+        repositories_users = database[:repositories_users]
+        repositories_users.delete
+        #logger.debug(entries[0])
+        # unique constraint failure when two syncs occur at once during spamming
+        repositories_users.import([:repository_id, :user_id], entries[0])
```

Comment:
It may be needed to run a query for every user. Then rewrite the import in a proper `INSERT INTO repositories_users (repository_id, user_id) SELECT repository.id, $UID FROM repositories WHERE name IN ($names) AND auth_required = true` (which import should do if the value is a query, as is done in `update_user_repositories`).